crypto/bp/bp_fp12.c

```diff
@@ -0,0 +1,881 @@
+ * Copyright 2016 MIRACL UK Ltd., All Rights Reserved. Portions of the
+ * attached software ("Contribution") are developed by MIRACL UK LTD., and
+ * are contributed to the OpenSSL project. The Contribution is licensed
+ * pursuant to the OpenSSL open source license provided above.
+ */
+
+#include "bp_lcl.h"
+
+FP12 *FP12_new()
+{
+    FP12 *ret = NULL;
+
+    if ((ret = OPENSSL_zalloc(sizeof(*ret))) == NULL)
+        return NULL;
+
+    ret->f[0] = FP6_new();
+    ret->f[1] = FP6_new();
+    if (ret->f[0] == NULL || ret->f[1] == NULL) {
+        FP6_free(ret->f[0]);
+        FP6_free(ret->f[1]);
+        return NULL;
+    }
+    return ret;
+}
+
+void FP12_clear(FP12 *a)
+{
+    FP6_clear(a->f[0]);
+    FP6_clear(a->f[1]);
+}
+
+void FP12_free(FP12 *a)
+{
+    if (a == NULL) {
+        return;
+    } else {
+        FP6_free(a->f[0]);
+        FP6_free(a->f[1]);
+    }
+    OPENSSL_free(a);
+}
+
+void FP12_clear_free(FP12 *a)
+{
+    if (a == NULL) {
+        return;
+    } else {
+        FP6_clear_free(a->f[0]);
+        FP6_clear_free(a->f[1]);
+    }
+    OPENSSL_free(a);
+}
+
+int FP12_rand(const BP_GROUP *group, FP12 *a)
+{
+    if (!FP6_rand(group, a->f[0]))
+        return 0;
+    if (!FP6_rand(group, a->f[1]))
+        return 0;
+    return 1;
+}
+
+void FP12_print(const FP12 *a)
+{
+    FP6_print(a->f[0]);
+    FP6_print(a->f[1]);
+}
+
+int FP12_zero(FP12 *a)
+{
+    if (!FP6_zero(a->f[0]))
+        return 0;
+    if (!FP6_zero(a->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP12_cmp(const FP12 *a, const FP12 *b)
+{
+    if (FP6_cmp(a->f[0], b->f[0]) != 0)
+        return 1;
+    if (FP6_cmp(a->f[1], b->f[1]) != 0)
+        return 1;
+    return 0;
+}
+
+int FP12_copy(FP12 *a, const FP12 *b)
+{
+    if (!FP6_copy(a->f[0], b->f[0]))
+        return 0;
+    if (!FP6_copy(a->f[1], b->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP12_is_zero(const FP12 *a)
+{
+    return FP6_is_zero(a->f[0]) & FP6_is_zero(a->f[1]);
+}
+
+int FP12_add(const BP_GROUP *group, FP12 *r, const FP12 *a, const FP12 *b)
+{
+    if (!FP6_add(group, r->f[0], a->f[0], b->f[0]))
+        return 0;
+    if (!FP6_add(group, r->f[1], a->f[1], b->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP12_sub(const BP_GROUP *group, FP12 *r, const FP12 *a, const FP12 *b)
+{
+    if (!FP6_sub(group, r->f[0], a->f[0], b->f[0]))
+        return 0;
+    if (!FP6_sub(group, r->f[1], a->f[1], b->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP12_neg(const BP_GROUP *group, FP12 *r, const FP12 *a)
+{
+    if (!FP6_neg(group, r->f[0], a->f[0]))
+        return 0;
+    if (!FP6_neg(group, r->f[1], a->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP12_mul(const BP_GROUP *group, FP12 *r, const FP12 *a, const FP12 *b,
+             BN_CTX *ctx)
+{
+    FP6 *t0 = NULL, *t1 = NULL, *t2 = NULL;
+    int ret = 0;
+
+    if ((t0 = FP6_new()) == NULL || (t1 = FP6_new()) == NULL
+        || (t2 = FP6_new()) == NULL)
+        goto err;
```

Comment:
`{}` around the block if the `if`-condition spans multiple lines. (I think this isn't explicit in our coding style, but it's easier to read.)